text/3700-ptr-assume-moved.md

```diff
@@ -0,0 +1,257 @@
+move has occurred is unfortunate in that it makes the support we add to the language more
+restrictive than the actual hardware reality allows for, but this seems to be the only solution
+available for the time being as modifying the entire stack to support disregarding the top bits of a
+pointer would be a non-trivial endeavour.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Without having a dedicated library helper for modifying the address, users wanting to make use of
+high-bit tagging would have to resort to manually using bitwise operations and would be at risk of
+inadvertently introducing Undefined Behaviour.  Having a helper for doing so in the library creates
+a place where e.g. Miri hooks can be called to let Miri know that a pointer's cannonical address has
+been updated. It also makes it possible for the helper to receive special treatment in the codegen
+stage, which is required for it to have the desired behaviour.
+
+It is most likely not feasible to make `assume_moved()` safe to use regardless of the context,
+hence the current approach is to make it an unsafe function with a safety notice about the user's
+responsibilities.
+
+# Prior art
+[prior-art]: #prior-art
+
+TBI already works in C, though mostly by default and care must be taken to make sure no Undefined
+Behaviour is introduced. The compiler does not take special steps to preserve the tags, but it
+doesn't try to remove them either.  That being said, the C/C++ standard library does not take
+tagging schemes into account during alias analysis. With this proposal, Rust would have much better
+defined and safer support for TBI than C or C++.
+
+Notably, [Android](https://source.android.com/docs/security/test/tagged-pointers) already makes
+extensive use of TBI by tagging all heap allocations.
+
+The idea is also not one specific to AArch64, as there are similar extensions present on other
+architectures that facilitate working with tagged pointers.
+This proposal is in no way AArch64-specific. As proposed, `assume_moved` could be used on x86
+for Intel's LAM or AMD's UAI as well. It could even be used for applications other than pointer
+tagging, such as using `mmap` to map the same allocation to two adjacent addresses and switch between
+them by using `assume_moved()`. Although that use-case requires its own documentation and
+additional concurrency safety considerations that do not apply to pointer tagging.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+What exactly should the helper be called? The main candidates so far have been `simulate_realloc`,
+`simulate_move` and `assume_moved`, alongside some other suggestions, but preferences seem to be
+rather subjective and no clearly most suitable name has emerged as of yet. It can be renamed to
+whichever name the relevant Rust team finds most appropriate.
```

Comment:
Let's please move discussion of the name to the rationale and alternatives section.  I'd prefer we make a presumptive decision on this at the RFC step and not leave it as an unanswered question.